cmd/kubeadm/app/phases/certs/certs.go

```diff
@@ -68,35 +69,74 @@ func CreatePKIAssets(cfg *kubeadmapi.MasterConfiguration, pkiPath string) (*x509
-	}
+	// Try to load ca.crt and ca.key from the PKI directory
+	caCert, caKey, err := pkiutil.TryLoadCertAndKeyFromDisk(pkiDir, certconstants.CACertAndKeyBaseName)
 
-	if err := writeKeysAndCert(pkiPath, "ca", caKey, caCert); err != nil {
-		return nil, fmt.Errorf("failure while saving CA keys and certificate [%v]", err)
-	}
-	fmt.Println("[certificates] Generated Certificate Authority key and certificate.")
+	if err == nil && caCert != nil && caCert.IsCA && caKey != nil {
+		fmt.Println("[certificates] Using the already-existing CA certificate and key.")
+	} else {
+		// An error occured while reading the certificate and the key (might simply not exist). That means we should generate our own CA.
+		caCert, caKey, err = pkiutil.NewCertificateAuthority()
+		if err != nil {
+			return nil, fmt.Errorf("failure while creating CA keys and certificate [%v]", err)
+		}
 
-	apiKey, apiCert, err := newServerKeyAndCert(caCert, caKey, altNames)
-	if err != nil {
-		return nil, fmt.Errorf("failure while creating API server keys and certificate [%v]", err)
+		if err = pkiutil.WriteCertAndKey(pkiDir, certconstants.CACertAndKeyBaseName, caCert, caKey); err != nil {
+			return nil, fmt.Errorf("failure while saving CA keys and certificate [%v]", err)
+		}
+		fmt.Println("[certificates] Generated Certificate Authority key and certificate.")
 	}
 
-	if err := writeKeysAndCert(pkiPath, "apiserver", apiKey, apiCert); err != nil {
-		return nil, fmt.Errorf("failure while saving API server keys and certificate [%v]", err)
+	// Try to load ca.crt and ca.key from the PKI directory
+	apiCert, apiKey, err := pkiutil.TryLoadCertAndKeyFromDisk(pkiDir, certconstants.APIServerCertAndKeyBaseName)
+
+	if err == nil && apiCert != nil && apiKey != nil && checkAltNamesExist(apiCert.IPAddresses, apiCert.DNSNames, altNames) {
+		fmt.Println("[certificates] Using the already-existing apiserver certificate and key.")
+	} else {
+		// An error occured while reading the certificate and the key (might simply not exist). That means we should generate our own certificate.
+		apiCert, apiKey, err := pkiutil.NewServerKeyAndCert(caCert, caKey, altNames)
+		if err != nil {
+			return nil, fmt.Errorf("failure while creating API server keys and certificate [%v]", err)
+		}
+
+		if err := pkiutil.WriteCertAndKey(pkiDir, certconstants.APIServerCertAndKeyBaseName, apiCert, apiKey); err != nil {
+			return nil, fmt.Errorf("failure while saving API server keys and certificate [%v]", err)
+		}
+		fmt.Println("[certificates] Generated API Server key and certificate")
 	}
-	fmt.Println("[certificates] Generated API Server key and certificate")
 
-	// Generate a private key for service accounts
-	saKey, err := certutil.NewPrivateKey()
-	if err != nil {
-		return nil, fmt.Errorf("failure while creating service account signing keys [%v]", err)
+	fmt.Printf("[certificates] Valid certificates and keys now exist in %q\n", pkiDir)
+
+	// TODO: Public phase methods should not return anything.
```

Comment:
Oh, yea. I meant no values like the CA cert that we're returning now. I'll update that